src/redis_ai_objects/script.c

```diff
@@ -197,30 +120,63 @@ int RedisAI_ScriptRun_IsKeysPositionRequest_ReportKeys(RedisModuleCtx *ctx,
     return REDISMODULE_OK;
 }
 
-RedisModuleType *RAI_ScriptRedisType(void) { return RedisAI_ScriptType; }
-
-int RAI_ScriptRunAsync(RAI_ScriptRunCtx *sctx, RAI_OnFinishCB ScriptAsyncFinish,
-                       void *private_data) {
-
-    RedisAI_RunInfo *rinfo = NULL;
-    RAI_InitRunInfo(&rinfo);
-
-    rinfo->single_op_dag = 1;
-    rinfo->OnFinish = (RedisAI_OnFinishCB)ScriptAsyncFinish;
-    rinfo->private_data = private_data;
-
-    RAI_DagOp *op;
-    RAI_InitDagOp(&op);
-
-    op->commandType = REDISAI_DAG_CMD_SCRIPTRUN;
-    op->devicestr = sctx->script->devicestr;
-    op->sctx = sctx;
-
-    rinfo->dagOps = array_append(rinfo->dagOps, op);
-    rinfo->dagOpCount = 1;
-    if (DAG_InsertDAGToQueue(rinfo) != REDISMODULE_OK) {
-        RAI_FreeRunInfo(rinfo);
+int RedisAI_ScriptExecute_IsKeysPositionRequest_ReportKeys(RedisModuleCtx *ctx,
+                                                           RedisModuleString **argv, int argc) {
+    // AI.SCRIPTEXECUTE script_name func KEYS n key....
+    if (argc < 6) {
         return REDISMODULE_ERR;
     }
-    return REDISMODULE_OK;
+    RedisModule_KeyAtPos(ctx, 1);
+    size_t argpos = 3;
+    if (argpos >= argc) {
+        return REDISMODULE_ERR;
+    }
+    long long count;
+    while (argpos < argc) {
+        const char *str = RedisModule_StringPtrLen(argv[argpos++], NULL);
+
+        // Inputs, outpus, keys, lists.
+        if ((!strcasecmp(str, "INPUTS")) || (!strcasecmp(str, "OUTPUTS")) ||
+            (!strcasecmp(str, "LIST_INPUTS")) || (!strcasecmp(str, "KEYS"))) {
+            bool updateKeyAtPos = false;
+            // The only scope where the inputs strings are 100% keys are in the KEYS and OUTPUTS
+            // scopes.
+            if ((!strcasecmp(str, "OUTPUTS")) || (!strcasecmp(str, "KEYS"))) {
+                updateKeyAtPos = true;
+            }
+            if (argpos >= argc) {
+                return REDISMODULE_ERR;
+            }
+            if (RedisModule_StringToLongLong(argv[argpos++], &count) != REDISMODULE_OK) {
+                return REDISMODULE_ERR;
+            }
+            if (count <= 0) {
+                return REDISMODULE_ERR;
+            }
+            if (argpos + count >= argc) {
+                return REDISMODULE_ERR;
+            }
+            for (long long i = 0; i < count; i++) {
+                if (updateKeyAtPos) {
+                    RedisModule_KeyAtPos(ctx, argpos);
+                }
+                argpos++;
+            }
+            continue;
+        }
+        // Timeout
```

Comment:
here we are checking for a `TIMEOUT` scope
there are two possible wrong scenarios
`AI.SCRIPTEXECUTE script func TIMEOUT 1 KEYS.....`
or
`AI.SCRIPTEXECUTE script func KEYS..... TIMEOUT" with no value
we will break the loop
and check if argpos==to argc. If not, it will fail